csrc/lower_thread_predicate.cpp

```diff
@@ -494,6 +519,185 @@ class RedundantUseAnalysis : BackwardVisitor {
 
 } // namespace
 
+// This function removes the redundant write to gmem when an output tensor has a
+// leaf domain merged from concretized broadcast root domain and parallelized by
+// thread/block id. issue https://github.com/csarofeen/pytorch/issues/2125
+void ThreadPredicateMap::avoidConcretizedBroadcastRedundantWrite(
+    const TensorView* out_tv) {
+  // step-1, check if there is a leaf domain parallelized by thread/block id and
+  // its definition is a merged
+  bool merged_parallelized_thread_block = false;
+  for (auto ld : out_tv->domain()->domain()) {
+    const ParallelType& pt = ld->getParallelType();
+    auto merge = dynamic_cast<Merge*>(ld->definition());
```

Comment:
Does this have to be a recursive merge? i.e. all expressions between root domains and this leaf domain must all be merge?

Reply:
Yes. This is the current assumption. If there is a split in the path, predicate away the indexing across the broadcast dimension becomes difficult.

Reply:
Or are you just saying it's the final expression you want to improve the write predication on?

Reply:
I meant sequential merges, so a merge operation going into another. Something along the lines of `for(auto expr : StmtSort::getExprs(fusion, ld)){...`

Reply:
Sorry, it should actually be similar to: `auto all_exp = DependencyCheck::getAllExprsBetween(
        {rfactor_domain.begin(), rfactor_domain.end()}, {ld});` as it should be bounded by the allocated domain.